src/platform/w3dengine/client/w3ddisplay.cpp

```diff
@@ -933,9 +1541,740 @@ bool W3DDisplay::Is_LetterBox_Fading()
+
+            sprintf(buf, "%.2fms ", time_for_frame * 1000.0f);
+            OutputDebugString(buf);
+
+            if (i >= 5) {
+                frame_time = frame_time + time_for_frame;
+                if (i > 6 && 2.0 * max_time_limit < time_for_frame / (float)(i - 4)) {
+                    i++;
+                    break;
+                }
+            }
+        }
+
+        frame_time = frame_time / (float)(i - 5);
+        count++;
+        sprintf(buf, "\n LOD %d, time %.2fms\n", lod, frame_time * 1000.0f);
+        OutputDebugString(buf);
+
+        if (frame_time < (float)max_time_limit && good_lod < lod) {
+            good_lod = lod;
+        }
+
+    } while (frame_time >= (float)max_time_limit && count < 10);
+
+    g_theWriteableGlobalData->m_terrainLOD = good_lod;
+    s_3DScene->Draw_Terrain_Only(false);
+    g_theTerrainRenderObject->Adjust_Terrain_LOD(0);
+#endif
+}
+
+void W3DDisplay::Render_LetterBox(unsigned int current_time)
+{
+    if (m_letterBoxEnabled) {
+        if (m_letterBoxFadeLevel != 1.0f) {
+            m_letterBoxFadeLevel = (float)(current_time - m_letterBoxFadeStartTime) / 1000.0f;
+
+            if (m_letterBoxFadeLevel > 1.0f) {
+                m_letterBoxFadeLevel = 1.0f;
+            }
+
+            unsigned int color = (unsigned int)(m_letterBoxFadeLevel * 255.0f) << 24;
+            Draw_Fill_Rect(0, 0, m_width, (int)(((float)m_height - (float)m_width * 0.5625f) * 0.5f), color);
+            Draw_Fill_Rect(
+                0, (int)((float)m_height - (((float)m_height - (float)m_width * 0.5625f) * 0.5f)), m_width, m_height, color);
+        }
+    } else if (m_letterBoxFadeLevel == 0.0f) {
+        m_letterBoxEnabled = 0;
+    } else {
+        m_letterBoxFadeLevel = 1.0f - (float)(current_time - m_letterBoxFadeStartTime) / 1000.0f;
+
+        if (m_letterBoxFadeLevel < 0.0f) {
+            m_letterBoxFadeLevel = 0.0f;
+        }
+
+        Draw_Fill_Rect(0,
+            0,
+            m_width,
+            (unsigned int)(((float)m_height - (float)m_width * 0.5625f) * 0.5f),
+            (unsigned int)(m_letterBoxFadeLevel * 255.0f) << 24);
+    }
+}
+
+void Reset_D3D_Device(bool restore_assets)
+{
+#ifdef BUILD_WITH_D3D8
+    if (g_theDisplay) {
+        if (W3D::Is_Initted()) {
+            if (!g_theDisplay->Get_Windowed()) {
+                if (restore_assets) {
+                    W3D::Set_Render_Device(W3D::Get_Render_Device(),
+                        g_theDisplay->Get_Width(),
+                        g_theDisplay->Get_Height(),
+                        g_theDisplay->Get_Bit_Depth(),
+                        g_theDisplay->Get_Windowed(),
+                        true,
+                        true,
+                        true);
+                    OSVERSIONINFOA VersionInformation;
+
+                    if (GetVersionEx(&VersionInformation)) {
+                        if (VersionInformation.dwPlatformId == VER_PLATFORM_WIN32_WINDOWS) {
```

Comment:
Why does it do this platform test here?

Reply:
No clue, I just copied what the original code does.